mixer/pkg/runtime2/handler/table.go

```diff
@@ -0,0 +1,141 @@
+
+	// Adapter that was used to create this Entry.
+	Adapter *adapter.Info
+
+	// Signature of the configuration used to create this entry.
+	Signature signature
+}
+
+// NewTable returns a new table, based on the given config snapshot. The table will re-use existing handlers as much as
+// possible from the old table.
+func NewTable(old *Table, snapshot *config.Snapshot, gp *pool.GoroutinePool) *Table {
+	t := &Table{
+		entries:  make(map[string]Entry),
+		counters: newTableCounters(snapshot.ID),
+	}
+
+	var f *factory
+
+	// Find all handlers, as referenced by instances, and associate to handlers.
+	instancesByHandler := config.GetInstancesGroupedByHandlers(snapshot)
+	for handler, instances := range instancesByHandler {
+		sig := calculateSignature(handler, instances)
+
+		currentEntry, found := old.entries[handler.Name]
+		if found && currentEntry.Signature.equals(sig) {
+			// reuse the Handler
+			t.entries[handler.Name] = currentEntry
+			t.counters.reusedHandlers.Inc()
+			continue
+		}
+
+		// instantiate the new Handler
+		if f == nil {
+			f = newFactory(snapshot)
+		}
+
+		e := NewEnv(snapshot.ID, handler.Name, gp)
+		instantiatedHandler, err := f.build(handler, instances, e)
+
+		if err != nil {
+			t.counters.buildFailure.Inc()
+
+			log.Errorf(
+				"Unable to initialize adapter: snapshot='%d', handler='%s', adapter='%s', err='%s'.\n"+
+					"Please remove the handler or fix the configuration.",
+				snapshot.ID, handler.Name, handler.Adapter.Name, err.Error())
+
+			continue
+		}
+
+		t.counters.newHandlers.Inc()
+
+		t.entries[handler.Name] = Entry{
+			Name:      handler.Name,
+			Handler:   instantiatedHandler,
+			Adapter:   handler.Adapter,
+			Signature: sig,
+		}
+	}
+
+	return t
+}
+
+// Cleanup the old table by selectively closing handlers that are not used in the given table.
+// The cleanup is called on the old table, based on a new config. To perf tableCounters on the current table
+// must be used.
+func (t *Table) Cleanup(current *Table) {
+	var toCleanup []Entry
+
+	for name, oldEntry := range t.entries {
+		if currentEntry, found := current.entries[name]; found && currentEntry.Signature.equals(oldEntry.Signature) {
+			// this entry is still in use. Skip it.
+			continue
+		}
+
+		// schedule for cleanup
+		toCleanup = append(toCleanup, oldEntry)
+	}
+
+	for _, entry := range toCleanup {
+		log.Debugf("Closing adapter %s/%v", entry.Name, entry.Handler)
+		current.counters.closedHandlers.Inc()
+		err := entry.Handler.Close()
+		if err != nil {
+			current.counters.closeFailure.Inc()
+			log.Warnf("Error closing adapter: %s/%v: '%v'", entry.Name, entry.Handler, err)
+		}
+	}
+}
+
+// Get returns the entry for a Handler with the given name, if it exists.
+func (t *Table) Get(handlerName string) (Entry, bool) {
+	e, found := t.entries[handlerName]
+	if !found {
+		return Entry{}, false
+	}
+
+	return e, true
+}
+
+// Empty returns an empty table instance.
+func Empty() *Table {
+	return &Table{}
```

Comment:
Is this is used frequently, you should store a &Table{} in a global variable and return that instead of a fresh instance every time.